### PR TITLE
Refactor test fixtures

### PR DESCRIPTION
**Pull request by Amertz08**
This PR refactors common test code into pytest fixtures for easy of reuse.

### CODE REVIEW

File: tests/test_processor.py [hunk 1]
```diff
@@ -169,7 +169,7 @@ def test_get_nameid_email_valid(self):
     def test_get_nameid_email_invalid(self):
         user = User(username='Test Name', email='test@email.com')
 
-        with pytest.raises(Exception, message=f"user_id {user.username} does not contain the '@' symbol, so is not a valid NameID Email address format."):
+        with pytest.raises(Exception, match=f"user_id {user.username} does not contain the '@' symbol, so is not a valid NameID Email address format."):
```
Comment (Amertz08) [inline tests/test_processor.py]: This test was failing as `message` isn't an argument.

**Comment by Amertz08** [comment]
@mhindery you should be able to enable Travis CI checks on the repo so that no PRs can be merged unless they pass the Travis builds.

File: conftest.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+@pytest.fixture()
+@lru_cache()
+def saml_request_minimal() -> str:
+    with (XML_ROOT / "min/request/sample_saml_request_minimal.xml").open("r") as f:
+        return f.read()
+
+
+@pytest.fixture()
+@lru_cache()
```
Comment (Amertz08) [inline conftest.py]: This caches the string being returned so that the fixture doesn't open the file every time it is being used.

**Comment by Amertz08** [comment]
Just doing a follow up on this.

File: djangosaml2idp/views.py [hunk 3]
```diff
@@ -143,16 +145,26 @@ def render_login_html_to_string(self, context=None, request=None, using=None):
...
-            except (TemplateDoesNotExist, TemplateSyntaxError) as e:
-                logger.error('Specified template {} cannot be used due to: {}. Falling back to default login template {}'.format(custom_login_template_name, str(e), default_login_template_name))
-                template = get_template(default_login_template_name, using=using)
-        else:
-            template = get_template(default_login_template_name, using=using)
+            template = self._fetch_custom_template(custom_login_template_name, default_login_template_name, using)
+            return template.render(context, request)
+
+        template = get_template(default_login_template_name, using=using)
         return template.render(context, request)
```
Comment (Amertz08) [inline djangosaml2idp/views.py:154]: There was no test coverage for the `render_login_html_to_string` method so I added some and then refactored for clarity.

File: djangosaml2idp/models.py [hunk 4]
```diff
@@ -50,6 +50,10 @@ def refresh_metadata(self, force_refresh: bool = False) -> bool:
         ''' If a remote metadata url is set, fetch new metadata if the locally cached one is expired. Returns True if new metadata was set.
             Sets metadata fields on instance, but does not save to db. If force_refresh = True, the metadata will be refreshed regardless of the currently cached version validity timestamp.
         '''
+
+        # TODO: this logic appears to be incorrect.
+        #  In the case local_metadata is Falsy and remote metadata is not set you error on extract
+        #  as it has nothing to extract on.
```
Comment (Amertz08) [inline djangosaml2idp/models.py:56]: @mhindery I added this note. This logic appears to be incorrect. I was adding test coverage for the models and ran across it.

Comment (mhindery) [reply djangosaml2idp/models.py:56]: Thanks for reporting this. I'll have a look at it, there are some save & validation cases in the admin which were previously causing some issues here so I'll check if it behaves nicely.

**Comment by mhindery** [comment]
On the topic of Travis CI, I've requested this to someone from the team as I cannot enable this for the organisation myself. I'll also have a look at the Github actions, which offer free ci testing as well.

**Comment by Amertz08** [comment]
Sounds good @mhindery. Also consider picking up a copy of TDD with Python. One of the best books on the subject for python. It's available for free online too.
https://www.obeythetestinggoat.com/

File: djangosaml2idp/models.py (hunk 4 above)
Comment (Amertz08) [reply djangosaml2idp/models.py:56]: Yeah if you return it back to it's original state you can see the test fails. It's currently xpasses. That being said it seemed like if that is the correct fix that there are 2 places noted that it seems like we should be returning false.

**Pull request merged**